spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageLocation.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.storage;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import org.springframework.util.Assert;
+
+/**
+ * Represents a Google Cloud Storage location provided by the user.
+ *
+ * @author Daniel Zou
+ */
+public final class GoogleStorageLocation {
+
+	private static final String GCS_URI_FORMAT = "gs://%s/%s";
+
+	private final String bucketName;
+
+	private final String blobName;
+
+	private final URI uri;
+
+	/**
+	 * Constructs a {@link GoogleStorageLocation} based on the provided Google Storage URI string.
+	 * The URI string is of the form: {@code gs://<BUCKET_NAME>/<PATH_TO_FILE>}
+	 *
+	 * @param gcsLocationUriString a Google Storage URI string to a bucket/folder/file.
+	 */
+	public GoogleStorageLocation(String gcsLocationUriString) {
+		try {
+			URI locationUri = new URI(gcsLocationUriString);
+
+			Assert.isTrue(
+					gcsLocationUriString.startsWith("gs://"),
+					"A Google Storage URI must start with gs://");
+
+			Assert.isTrue(
+					locationUri.getAuthority() != null,
+					"No bucket specified in the location: " + gcsLocationUriString);
+
+			this.bucketName = locationUri.getAuthority();
+			this.blobName = getBlobPathFromUri(locationUri);
+
+			// ensure that if it's a bucket handle, location ends with a '/'
+			if (this.blobName == null && !gcsLocationUriString.endsWith("/")) {
+				locationUri = new URI(gcsLocationUriString + "/");
+			}
+			this.uri = locationUri;
+		}
+		catch (URISyntaxException e) {
+			throw new IllegalArgumentException(
+					"Invalid location: " + gcsLocationUriString);
+		}
+	}
+
+	/**
+	 * Returns the Google Storage bucket name.
+	 */
+	public String getBucketName() {
+		return bucketName;
+	}
+
+	/**
+	 * Returns the path to the blob/folder relative from the bucket root. Returns the empty
+	 * String if the {@link GoogleStorageLocation} specifies a bucket itself.
+	 */
+	public String getBlobName() {
+		return blobName;
+	}
+
+	/**
+	 * Returns the GCS URI of the location.
+	 */
+	public URI uri() {
+		return this.uri;
+	}
+
+	/**
+	 * Returns the Google Storage URI string for the location.
+	 */
+	public String uriString() {
+		String processedBlobName = (blobName == null) ? "" : blobName;
```

Comment:
The logic for returning an empty string for a null blobName seems useful both here and in `getBlobName()`. Perhaps `blobName` could be stored as an empty string instead of null in the first place? Are there any times when a `null` value is in itself useful?

Reply:
Agreed, I think empty string would be better too. I kept the `blobName` null because the `getBlobName` method in `GoogleStorageResource` returns null; I was worried if I changed it to empty string that I may accidentally break some user who expect it to return null.

@meltsufin - Thoughts on this? Would it be safe to change so that it returns the empty string?

Reply:
Blob name being null is an indication that it is a bucket resource (think directory). So, I think we should keep it that way.

Reply:
OK, will leave as null then. 

Just to clarify, is there a reason why the empty string could not be used to indicate a bucket resource?

Reply:
It probably can, but I'm not sure if an empty string is a valid filename... probably not. Null seems to be more semantically appropriate as the indicator.